File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/SumAggregationResult.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.searchlib.expression.SingleResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an aggregated result holding the sum of the aggregating expression for all matching hits.
 *
 * @author baldersheim
 * @author Simon Thoresen Hult
 */
public class SumAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 82, SumAggregationResult.class);
    private SingleResultNode sum;

    /**
     * Constructs an empty result node. <b>NOTE:</b> This instance is broken until non-optional member data is set.
     */
    public SumAggregationResult() {

    }

    /**
     * Constructs an instance of this class with given sum.
     *
     * @param sum The initial sum to set.
     */
    public SumAggregationResult(SingleResultNode sum) {
        setSum(sum);
    }

    /**
     * Returns the sum of all results in this.
     *
     * @return The numeric sum.
     */
    public final SingleResultNode getSum() {
        return sum;
    }

    /**
     * Sets the sum of all results in this.
     *
     * @param sum The sum to set.
     * @return This, to allow chaining.
     */
    public final SumAggregationResult setSum(SingleResultNode sum) {
        this.sum = sum;
        return this;
    }

    @Override
    public ResultNode getRank() {
        return sum;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        serializeOptional(buf, sum);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        sum = (SingleResultNode)deserializeOptional(buf);
    }

    @Override
    protected void onMerge(AggregationResult result) {
        sum.add(((SumAggregationResult)result).sum);
    }

    @Override
    public SumAggregationResult clone() {
        SumAggregationResult obj = (SumAggregationResult)super.clone();
        if (sum != null) {
            obj.sum = (SingleResultNode)sum.clone();
        }
        return obj;
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        return equals(sum, ((SumAggregationResult)obj).sum);
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("sum", sum);
    }
}

File: searchlib/src/vespa/searchlib/aggregation/sumaggregationresult.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "sumaggregationresult.h"

#include <vespa/document/fieldvalue/document.h>
#include <vespa/searchlib/expression/floatresultnode.h>
#include <vespa/searchlib/expression/resultvector.h>

#include <vespa/searchlib/aggregation/aggregation.hpp>
#include <vespa/vespalib/objects/visit.hpp>

using search::expression::FloatResultNode;
using search::expression::NumericResultNode;
using search::expression::ResultNodeVector;
using vespalib::Deserializer;
using vespalib::Serializer;

namespace search::aggregation {

IMPLEMENT_AGGREGATIONRESULT(SumAggregationResult, AggregationResult);

SumAggregationResult::SumAggregationResult() : AggregationResult(), _sum(FloatResultNode(0.0)) {
}

SumAggregationResult::SumAggregationResult(NumericResultNode::UP sum) : AggregationResult(), _sum(sum.release()) {
}

SumAggregationResult::~SumAggregationResult() = default;

void SumAggregationResult::onPrepare(const ResultNode& result) {
    if (!is_ready(_sum.get(), result)) {
        _sum = create_and_ensure_wanted<NumericResultNode, FloatResultNode>(result);
    }
}

void SumAggregationResult::initForUnitTest(const ResultNode& result) {
    onPrepare(result);
    _sum->set(result);
}

void SumAggregationResult::onMerge(const AggregationResult& b) {
    _sum->add(*static_cast<const SumAggregationResult&>(b)._sum);
}

void SumAggregationResult::onAggregate(const ResultNode& result) {
    if (result.isMultiValue()) {
        static_cast<const ResultNodeVector&>(result).flattenSum(*_sum);
    } else {
        _sum->add(result);
    }
}

void SumAggregationResult::onReset() {
    _sum.reset(static_cast<NumericResultNode*>(_sum->getClass().create()));
}

Serializer& SumAggregationResult::onSerialize(Serializer& os) const {
    AggregationResult::onSerialize(os);
    return os << _sum;
}

Deserializer& SumAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    return is >> _sum;
}

void SumAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "sum", _sum);
}

} // namespace search::aggregation
